### PR TITLE
Import the wpcom package to the monorepo

### PR DESCRIPTION
**Pull request by jsnajdr**
**Already done:**
- import the whole revision history of https://github.com/Automattic/wpcom.js with `lerna import`
- organize the sources by Calypso conventions: `src/` folder, building with `calypso-build`
- make Calypso use the monorepo package instead of the NPM-published one. Needed a fix when importing the module as ESM.
- the `wpcom` library is also published as a single script file to be used in a "traditional" way with a global variable:
```
<script src="wpcom.js">
<script>
  const wpcom = WPCOM( token );
  wpcom.site( ... );
</script>
```
That script uses webpack to work. Made this a step in the `prepare` script.

**Todo:**
- make tests work. The tests are not unit tests, i.e., running in complete isolation, but rather integration tests that run against a production WordPress.com REST API as authenticated user. Not suitable for CI, but good for running locally.
- there is the `webapp/` dir that I think is used to produce content for http://wpcomjs.com/
- there is `examples/` dir with example apps. We need to ensure they work.

### CODE REVIEW

**Comment by jsnajdr** [comment]
A possible strategy how to deliver this project is:
1. Import the `wpcom` repo, with all the history, as a merge commit of a standalone PR. And keep it just "lying around" in the `wp-calypso` repo.
2. Adapt and activate the package in a followup PR, or possibly multiple ones.

File: packages/wpcom.js/.gitignore
Comment (sgomes) [inline packages/wpcom.js/.gitignore]: Should we really include this file?

File: packages/wpcom.js/Makefile
Comment (sgomes) [inline packages/wpcom.js/Makefile]: Do we need this file at all, if we're switching to `calypso-build`?

File: packages/wpcom.js/.gitignore
Comment (sgomes) [inline packages/wpcom.js/.gitignore]: Should we really include this file?

File: packages/wpcom.js/examples/browser-cors/package.json
Comment (sgomes) [inline packages/wpcom.js/examples/browser-cors/package.json]: I'd prefer to drop the package files for examples; there's not much point in them.

File: packages/wpcom.js/examples/media-editor/webpack.config.js
Comment (sgomes) [inline packages/wpcom.js/examples/media-editor/webpack.config.js]: How about dropping the build files for the examples? We're not going to be shipping them, and I think they're mostly useful as just code samples anyway. Keeping these targets up and running out of the box is going to add maintenance overhead as webpack releases new major versions, for example.

File: packages/wpcom.js/examples/server/package.json
Comment (sgomes) [inline packages/wpcom.js/examples/server/package.json]: Same as mentioned for other example package file.

File: packages/wpcom.js/package.json
Comment (sgomes) [inline packages/wpcom.js/package.json]: Isn't this technically a breaking change, since the paths are changing inside the package?

File: packages/wpcom.js/package.json
Comment (sgomes) [inline packages/wpcom.js/package.json]: I'm not sure if we need to do this in the `cjs` version, but it's probably safe to do so.

**Review by sgomes — COMMENTED** [review]
Thank you for your great work on this, @jsnajdr! 👍 

In general, I like the approach of importing the whole thing as-is first, before activating it. We just have to be very careful about making sure we don't publish it in the meantime.

Besides the in-code comments, I  wonder if it's worth importing the tests as they are at all? It's all done with mocha-based E2E tests, which don't fit very well with the way the monorepo does things. If we are to use them as they are, we'll need to add new dev dependencies, as well as a new build target to the overall monorepo, just for this package.

File: packages/wpcom.js/package.json
Comment (jsnajdr) [reply packages/wpcom.js/package.json]: It depends on whether you consider the paths inside the package part of the public API, or an internal implementation detail 🙂 

I think it's safe to do a major update to 6.0.0, even though I'm not 100% sure if it's really necessary.

File: packages/wpcom.js/examples/media-editor/webpack.config.js
Comment (jsnajdr) [reply packages/wpcom.js/examples/media-editor/webpack.config.js]: That would mean that we provide some example code that you can study, but you need to BYOBS (bring your own build script) if you want to actually run it? Sounds good to me.

File: packages/wpcom.js/Makefile
Comment (jsnajdr) [reply packages/wpcom.js/Makefile]: The `Makefile` and `.gitignore` still contains some scripts information that needs to be converted to something else -- NPM build scripts in `package.json` and a better `.gitignore`. That's why I didn't delete them yet.

**Comment by jsnajdr** [comment]
> Besides the in-code comments, I wonder if it's worth importing the tests as they are at all?

Yes, I think we could save a lot of effort if we just drop them. They can't be easily automated, and I don't think many people will run then locally. The situation would be different for a library that's actively developed, but this one is a legacy library that we want to deprecate.

File: packages/wpcom.js/Makefile
Comment (sgomes) [reply packages/wpcom.js/Makefile]: Sounds good, let's keep them, then 👍

File: packages/wpcom.js/examples/media-editor/webpack.config.js
Comment (sgomes) [reply packages/wpcom.js/examples/media-editor/webpack.config.js]: Yes, sorry I wasn't very clear, that's exactly what I meant 🙂 And I love the acronym! 😄

File: packages/wpcom.js/package.json
Comment (sgomes) [reply packages/wpcom.js/package.json]: That's a good question; I assumed it was part of the public API, but I could see the argument being made for internal implementation detail. I'll leave it to your judgement. I assume we won't be releasing anything for now, and probably not before making other breaking changes?

File: packages/wpcom.js/.gitignore
Comment (sgomes) [reply packages/wpcom.js/.gitignore]: As per your comment above, let's keep this file until we convert the data here into something else.

**Comment by sirreal** [comment]
> * Not suitable for CI, but good for running locally.

We do run integration tests on a schedule. If we wanted to keep them, they could be added:

https://github.com/Automattic/wp-calypso/blob/1f7d696c5873309e80221590f34dcc12255b7fd4/.circleci/config.yml#L381-L401
https://github.com/Automattic/wp-calypso/blob/1f7d696c5873309e80221590f34dcc12255b7fd4/test/integration/jest.config.js#L1-L15

**Comment by sgomes** [comment]
> We do run integration tests on a schedule. If we wanted to keep them, they could be added:

These are browser-based mocha tests, however. Do we have any infrastructure in place for that sort of thing?

**Comment by sirreal** [comment]
> These are browser-based mocha tests, however. Do we have any infrastructure in place for that sort of thing?

Not to the best of my knowledge.

**Comment by jsnajdr** [comment]
The big import task was done separately in #39080. Now this branch contains only changes to the imported sources.

File: packages/wpcom.js/package.json
Comment (sgomes) [inline packages/wpcom.js/package.json]: I think we should preserve this for now. This is only one of several PRs that will be modernising this, right?

File: packages/wpcom.js/package.json
Comment (sgomes) [inline packages/wpcom.js/package.json]: Maybe best left for another PR, but we should update the version here to match the monorepo's.

File: packages/wpcom.js/package.json
Comment (sgomes) [inline packages/wpcom.js/package.json]: Should we delete the list of contributors? Other packages don't appear to have it, and it doesn't seem particularly useful.

File: packages/wpcom.js/package.json
Comment (jsnajdr) [reply packages/wpcom.js/package.json]: I think we need it both in `cjs` and in `esm` versions. It makes sure that `import './fs'` and `require( './fs' )` are resolved to `fs-browser.js` when running in browser.

File: packages/wpcom.js/package.json
Comment (jsnajdr) [reply packages/wpcom.js/package.json]: OK, I separated activating the package and using it in Calypso to the last commit in the branch. That way we can test it really works in Calypso and remove the commit before merging.

File: packages/wpcom.js/package.json
Comment (jsnajdr) [reply packages/wpcom.js/package.json]: You mean the version of `@babel/runtime` or `debug`? Or something else?

File: packages/wpcom.js/package.json
Comment (jsnajdr) [reply packages/wpcom.js/package.json]: Yes, deleted the list of contributors and leaving only "Author: Automattic Inc."

**Comment by jsnajdr** [comment]
This PR now contains the following set of changes:
- update build scripts to use `calypso-build` and build both modules and the all-in-one bundle
- rename the published `build/` directory to `dist/`
- clean up `package.json` fields
- use the package in Calypso app (only for testing, will be removed before merging and added back before all other wpcom.js PRs are merged)

**Comment by jsnajdr** [comment]
The original `wpcom` build that used `n8-make` also created sourcemaps for the transpiled files (e.g., `build/lib/site.comment.js.map`). In other words, it ran Babel with the `--source-maps` option. `calypso-build/transpile` doesn't do that and source maps are not generated and published. Does it matter?

File: packages/wpcom.js/package.json
Comment (sgomes) [reply packages/wpcom.js/package.json]: I meant `debug` specifically, but `@babel/runtime` too if they're out of sync.

Although, is `@babel/runtime` really needed as a direct dependency here, now that it's using `calypso-build`? Not entirely sure how that works.

File: packages/wpcom.js/package.json
Comment (sgomes) [reply packages/wpcom.js/package.json]: Just to clarify, if the goal is to use the library within the monorepo, I don't think it needs to be unmarked as `private`.

I'm ok with making it public at the end of this PR, but it doesn't have to be done, if e.g. there are other breaking changes planned for subsequent PRs. For example, I would like to drop the dependency on `wpcom-xhr-request`, since it's only used as a default request handler. I would much rather have the burden of choosing a request handler resting solely on the consumer.

Comment (jsnajdr) [reply packages/wpcom.js/package.json]: > Just to clarify, if the goal is to use the library within the monorepo, I don't think it needs to be unmarked as `private`.

That's true, I totally missed that! Putting the `private` field back.

File: packages/wpcom.js/package.json
Comment (jsnajdr) [reply packages/wpcom.js/package.json]: That's a good point, I updated versions of `debug` and `@babel/runtime` to the newer ones used by the monorepo.

`@babel/runtime` is still needed. The published modules (both CJS and ESM) are transpiled by Babel using the `@wordpress/browserslist-config`. And when transpiling class syntax, for example, the transpiled modules import a few helpers:
```js
import _classCallCheck from "@babel/runtime/helpers/classCallCheck";
import _createClass from "@babel/runtime/helpers/createClass";
```
That makes `@babel/runtime` a runtime dependency of the published package.

Comment (sgomes) [reply packages/wpcom.js/package.json]: That makes sense, thank you! 👍

**Review by sgomes — APPROVED** [review]
Thank you for all your great work on this, @jsnajdr! There's a lot of further work we can do to improve `wpcom` in the near future, but I think we're ready to switch to it in the monorepo with this PR 👍

**Comment by sgomes** [comment]
@jsnajdr FYI, this PR appears to come with the bonus of reduced bundle sizes 🙂

**Comment by jsnajdr** [comment]
> FYI, this PR appears to come with the bonus of reduced bundle sizes

That's great! It started to work only after my latest push, where I updated the lockfile to get rid of the NPM-published `wpcom@5` dependency and made sure that the local file reference is used.

**Comment by sirreal** [comment]
Nice work, thanks!

It would be great to add a deprecation message to the README and archive the repo now:
https://github.com/Automattic/wpcom.js

[example message](https://github.com/Automattic/calypso-codemods/commit/90f2b32fd4298ec5027292107825f3ffb5daac3a)

**Pull request merged**
